examples/benchmark/utils.py

```diff
@@ -0,0 +1,215 @@
+
+    Returns
+    -------
+    np.ndarray
+        Indices of detected heartbeats.
+    """
+    if detector == 'mne':
+        detection = mne.preprocessing.ecg.qrs_detector(fs, ecg, verbose=False)
+    elif detector == 'wfdb-xqrs':
+        detection = wfdb.processing.xqrs_detect(ecg, fs, verbose=False)
+    elif detector == 'pyecg-pantompkins':
+        detection = ecgdetectors.Detectors(fs).pan_tompkins_detector(ecg)
+    elif detector == 'biosppy-hamilton':
+        detection = biosppy.signals.ecg.hamilton_segmenter(ecg, fs)[0]
+    elif detector == 'heartpy':
+        wd, m = heartpy.process(ecg, fs)
+        detection = np.array(wd['peaklist'])[wd['binary_peaklist'].astype(bool)]
+    elif detector == 'neurokit2-nk':
+        clean_ecg = neurokit2.ecg.ecg_clean(ecg, int(fs), method='neurokit')
+        detection = neurokit2.ecg.ecg_findpeaks(clean_ecg, int(fs), method='neurokit')['ECG_R_Peaks']  # noqa
+    elif detector == 'neurokit2-kalidas2017':
+        clean_ecg = neurokit2.ecg.ecg_clean(ecg, int(fs), method='kalidas2017')
+        detection = neurokit2.ecg.ecg_findpeaks(clean_ecg, int(fs), method='kalidas2017')['ECG_R_Peaks']  # noqa
+    elif detector == 'sleepecg-c':
+        detection = sleepecg.detect_heartbeats(ecg, fs, backend='c')
+    elif detector == 'sleepecg-numba':
+        detection = sleepecg.detect_heartbeats(ecg, fs, backend='numba')
+    elif detector == 'sleepecg-python':
+        detection = sleepecg.detect_heartbeats(ecg, fs, backend='python')
+    else:
+        raise ValueError(f'Unknown QRS detector: {detector}')
+    return np.asarray(detection)
+
+
+def evaluate_single(
+    record: ECGRecord,
+    detector: str,
+    signal_len: int,
+    max_distance: float,
+    calc_rri_similarity: bool,
+    timeout: int,
+    max_timeouts: int,
+) -> Dict[str, Any]:
+    """
+    Evaluate a heartbeat detector on a given annotated ECG record.
+
+    Optionally, similarity measures between detected and annotated
+    RR intervals can be calculated. As this requires interpolation, it may
+    take some time for long signals. Setting a timeout only works on Unix
+    systems.
+
+    Parameters
+    ----------
+    record : ECGRecord
+        As received from `reader_dispatch`.
+    detector : str
+        String identifier of the detector to be used.
+    signal_len : int
+        Length to which the signal should be sliced.
+    max_distance : float
+        Maximum temporal distance in seconds between detected and annotated
+        beats to count as a successful detection.
+    calc_rri_similarity : bool
+        If `True`, calculate similarity measures between detected and
+        annotated RR-intervals (computationally expensive for long
+        signals).
+    timeout : int
+        Number of seconds after which to attempt cancelling execution using
+        SIGALRM.
+    max_timeouts : int
+        Number of timeouts after which a detector is skipped completely.
```

Comment:
I had messed the defaults in the readme, those are fixed now.

If a detector times out, it carries on with the next record, so it won't run forever.

`signal.alarm(timeout)` asks the OS for a `SIGALRM` after `timeout seconds`, which is then handled by `sigalrm_handler`, which in turn raises the `TimeoutError`. The triggering time limit is set in `config.yml` using the key `timeout`. 

We'd need to limit it to 300 minutes then (see "Known Issues" at the bottom of `README.md`), since `wfdb-xqrs` behaves strangely for some records. Or figure out the maximum signal length where it still works.

Reply:
> I had messed the defaults in the readme, those are fixed now.

👍 

> If a detector times out, it carries on with the next record, so it won't run forever.

This isn't entirely clear from reading the docs, but it now makes sense. The relationship between `timeout` and `max_timeouts` isn't immediately clear. Also, `timeout=0` is a special case which needs to be documented.

> `signal.alarm(timeout)` asks the OS for a `SIGALRM` after `timeout` seconds, which is then handled by `sigalrm_handler`, which in turn raises the `TimeoutError`. The triggering time limit is set in `config.yml` using the key `timeout`.

Got it. But all runs in the config file use the default of zero.

> We'd need to limit it to 300 minutes then (see "Known Issues" at the bottom of `README.md`), since `wfdb-xqrs` behaves strangely for some records. Or figure out the maximum signal length where it still works.

I'm tempted to do that. The whole timeout thing is just because this detector cannot handle long signals. Why don't we hard-code the max signal length for `wfdb-xqrs`? Then we could still use lengths up to 600 minutes for all other detectors.

Alternative suggestion: we could limit `timeout` and `max_timeouts` to values that would lead to skipping `wfdb-xqrs` for 600+ minutes, but not others.

Reply:
But given that canceling with `SIGALRM` doesn't work for this detector kind of means that this isn't really that useful.

Reply:
> Got it. But all runs in the config file use the default of zero.

Right, that is because apparently I was a bit too enthusiastic when cleaning up that file 🙈 

Maybe a summary of the issues will help here:
- `heartpy ` 
  - has extreme variability in it's runtime (e.g. for 600 signal minutes, runtimes are between 5 and 1135 seconds (not a typo))
  - above a signal length somewhere between 600 and 900 minutes, ltdb:15184:ECG2 takes at least several hours
  - can't be interrupted by `SIGALRM` (or at least the part taking so long can't be)
- `wfdb-xqrs`
  - mainly has low runtime variability
  - above a signal length somewhere between 300 and 600 minutes, ltdb:14134:ECG2 and ltdb:14184:ECG2 take much longer than the other records at the same length
  - can be interrupted by `SIGALRM`
- `biosppy-hamilton`
  - also has low runtime variability
  - doesn't scale well with signal length
  - can be interrupted by `SIGALRM`

So the simplest solution would likely be to use only signal lengths up to 300 minutes and add these notes to the readme for anyone who would want to try longer signals. This would result in the following plots:
![runtime__300](https://user-images.githubusercontent.com/22592497/130257452-489b92fc-381f-4124-9e6f-1d614ed66ac8.png)
![runtime_per_minute__300](https://user-images.githubusercontent.com/22592497/130257456-c7cd4100-ccab-4d66-a542-9d34639f0e7b.png)

Reply:
I think this is a good solution. We could even create another entry in the config (e.g. `runtime-long`), where we exclude these slow detectors and explain why. Then we can have another figure showing longer signal lengths only, which is particularly interesting because we're doing sleep ECG analysis, so those algorithms cannot be used for this purpose.

Reply:
But I think it will make the benchmark script much simpler if we get rid of the timeout and canceling business.

Reply:
> I think this is a good solution. We could even create another entry in the config (e.g. `runtime-long`), where we exclude these slow detectors and explain why. Then we can have another figure showing longer signal lengths only, which is particularly interesting because we're doing sleep ECG analysis, so those algorithms cannot be used for this purpose.

Since all detectors which would be included in a `runtime-long` benchmark scale more or less linearly with signal length, I think we could leave it at the 300 minute max. benchmark.

Reply:
OK, then let's make sure to mention which detectors cannot be used for signals longer than about five hours in the docs!

Reply:
The "strange behaviours" are mentioned in "Known Issues" in the benchmark readme and I think what runtime to expect from which detector should be clear from the plot. Would you like to add anything?

Reply:
Maybe mention that `heartpy` fails on one signal which we're skipping by catching the `BadSignalWarning`?